test/users/alvaradoo/refToSubclassInternalError-shared.chpl

```diff
@@ -0,0 +1,89 @@
+use BlockDist;
+
+/* Wrapper concrete class for generic class. */
+class GenProperty {
+  var dataType: int;
+}
+
+/* Wrapped generic class to hold arrays of variable size and type. */
+class Property: GenProperty {
+  type etype;
+  var propertyIdentifier: domain(int);
+  var propertyValue: [propertyIdentifier] etype;
+}
+
+/** Given input data from dataArray and the vertices (rows) to insert the data,
+* insert every key:value pair to the corresponding element of vertexProps. */
+proc populate(dataArray, inputVertices, ref vertexProps, dataTypeMap, colId) {
+  var etype = dataArray.eltType:string;
+  var etypeInd:int = dataTypeMap.find(etype);
+  writeln("## Inserting ", etype, " data ##");
+  forall (v,j) in zip(inputVertices,inputVertices.domain) {
+    writeln("Attempting to insert data ", dataArray[j], " for vertex ", inputVertices[j]);
+    ref currentProperty: shared Property = vertexProps[v,etypeInd];
```

Comment:
The type `shared Property` is invalid for `ref currentProperty` because the RHS has the type `shared GenProperty?`, which is not an instantiation of `shared Property`. The .good file needs to indicate a compilation error. See also #23644.

This also applies to the other test in this PR.

Reply:
Only because the array is made of the more general type, every instance in the array is actually of the subclass.

Reply:
However, it's definitely an improvement over the behavior before